`TestHoodieDeltaStreamerWithMultiWriter#testLatestCheckpointCarryOverWithMultipleWriters` is flaky

### Bug Description

**What happened:**
`TestHoodieDeltaStreamerWithMultiWriter#testLatestCheckpointCarryOverWithMultipleWriters` is flaky
```
[ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 152.572 s <<< FAILURE! - in org.apache.hudi.utilities.deltastreamer.TestHoodieDeltaStreamerWithMultiWriter
[ERROR] testLatestCheckpointCarryOverWithMultipleWriters{HoodieTableType}[1]  Time elapsed: 22.821 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <00008> but was: <00009>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1145)
	at org.apache.hudi.utilities.deltastreamer.TestHoodieDeltaStreamerWithMultiWriter.verifyCommitMetadataCheckpoint(TestHoodieDeltaStreamerWithMultiWriter.java:318)
	at org.apache.hudi.utilities.deltastreamer.TestHoodieDeltaStreamerWithMultiWriter.testLatestCheckpointCarryOverWithMultipleWriters(TestHoodieDeltaStreamerWithMultiWriter.java:310)
```

**What you expected:**
Test should pass each time.

**Steps to reproduce:**
1. Run Azure CI, or run the test locally until it fails


### Environment

**Hudi version:** master
**Query engine:** (Spark/Flink/Trino etc)
**Relevant configs:**


### Logs and Stack Trace

_No response_

--- a/hudi-utilities/src/test/java/org/apache/hudi/utilities/deltastreamer/TestHoodieDeltaStreamerWithMultiWriter.java
+++ b/hudi-utilities/src/test/java/org/apache/hudi/utilities/deltastreamer/TestHoodieDeltaStreamerWithMultiWriter.java
@@ -298,6 +298,12 @@ public class TestHoodieDeltaStreamerWithMultiWriter extends HoodieDeltaStreamerT
     meta.reloadActiveTimeline();
     int totalCommits = meta.getCommitsTimeline().filterCompletedInstants().countInstants();
 
+    // Capture the checkpoint after backfillJob.sync() to calculate expected final checkpoint
+    HoodieCommitMetadata commitMetadataAfterBackfill = getLatestMetadata(meta);
+    String checkpointAfterBackfill = CheckpointUtils.getCheckpoint(commitMetadataAfterBackfill).getCheckpointKey();
+    int expectedNextCheckpointNum = Integer.parseInt(checkpointAfterBackfill) + 1;
+    String expectedFinalCheckpoint = String.format("%05d", expectedNextCheckpointNum);
+
     // add a new commit to timeline which may not have the checkpoint in extra metadata
     addCommitToTimeline(meta);
     meta.reloadActiveTimeline();
@@ -307,7 +313,7 @@ public class TestHoodieDeltaStreamerWithMultiWriter extends HoodieDeltaStreamerT
     new HoodieDeltaStreamer(cfgBackfillJob, jsc).sync(); // if deltastreamer checkpoint fetch does not walk back to older commits, this sync will fail
     meta.reloadActiveTimeline();
     Assertions.assertEquals(totalCommits + 2, meta.getCommitsTimeline().filterCompletedInstants().countInstants());
-    verifyCommitMetadataCheckpoint(meta, "00008");
+    verifyCommitMetadataCheckpoint(meta, expectedFinalCheckpoint);
   }
 
   private void verifyCommitMetadataCheckpoint(HoodieTableMetaClient metaClient, String expectedCheckpoint) throws IOException {

